RFC - Deprecate duplicate methods in Router

markstory opened the pull request:
All of these methods have nicer/better cousins on `RouteBuilder`. By deprecating these methods we can remove them for 4.x. This is a possibly a contentious change as I suspect a few people are using these methods. I _really_ wanted to remove `Router::connect()` as well but I feel that might be too big of a change.

One of the reasons I want to remove these methods is they create broken use cases when combined with scopes. For example:

``` php
Router::scope('/api', function ($routes) {
  Router::mapResources('Articles');
});
```

Does not do what you want. There are many other ways to get into these bad states, and I'd prefer to not have them.

markstory committed:
Deprecate some of the static methods in Router.

All of these methods have nicer/better cousins on RouteBuilder. By
deprecating these methods we can remove them for 4.x. This is a possibly
a contentious change as I suspect a few people are using these methods.

markstory committed:
Fix up tests for deprecated methods.

I'm not a fan of messing around with the error_reporting level, but
PHPUnit doesn't give us many better options.

lorenzo commented:
This is going to take a lot of people by surprise, but good stuff!

markstory commented:
@lorenzo Should we not do this now? I really wanted to remove `Router::connect()` but I felt that might need more warning. If you think this is going to anger lots of folks we can hold off until a later 3.x release.

thinkingmedia commented on src/Routing/Router.php:
```diff
@@ -219,9 +219,14 @@ public static function connect($route, $defaults = [], $options = [])
      *   shifted into the passed arguments. As well as supplying patterns for routing parameters.
      * @return void
      * @see \Cake\Routing\RouteBuilder::redirect()
+     * @deprected 3.3.0 Use Router::scope() and RouteBuilder::redirect() instead.
```
should be ``deprecated``
missing the **a**

markstory replied:
Thank you.

markstory committed:
Fix typo.

thinkingmedia commented:
I'm not a fan of run-time errors for deprecated methods. The phpDoc notation is enough to flag the method in the IDE (I use PHPStorm). Anyone not using an IDE with phpDoc support is out of luck I guess.

markstory commented:
@thinkingmedia Runtime errors ensure that people can't say there wasn't warning. Would you recommend we do doc tags now and add runtime errors closer to 4.0? For example 3.5 or 3.6?

lorenzo commented:
Let's tag first, remove any reference to those methods in the docs, and in 3.4 we start triggering errors

jhli commented:
:+1: for notices at some point (I don't care when).

markstory commented:
@lorenzo I'm good with that.

markstory committed:
Remove runtime errors for router deprecations.

We'll start with documentation tags for now. In a future release runtime
errors will be added.
